### PR TITLE
SRE 1518 Show Circle CI deployment progress

### PR DESCRIPTION
**Pull request by paikattilrohin**
Closes https://snapsheettech.atlassian.net/browse/SRE-1518

## What
Currently there the triggered CircleCI job isn't monitored

## Why
This enables projects to monitor the triggered CircleCI Workflows 

## How
Use the CircleCI API to poll the status of the triggered Workflow and update the github action results accordingly.  

## Testing
Check out a branch of your project that uses this GitHub Action.

Update the action to reference this branch:
```yaml
uses:  snapsheet/trigger-circleci-pipeline-action@SRE-1518-show-circle-ci-deployment-progress
```

Check the logs of the github action to see that the CircleCI Job is monitored.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -94,6 +96,11 @@ axios
     setOutput("id", response.data.id);
     setOutput("number", response.data.number);
     setOutput("state", response.data.state);
+    workFlowUrl =
+      "https://circleci.com/api/v2/pipeline/" + response.data.id + "/workflow";
+    info(
```
Comment (whatisdot) [inline index.js]: Can you test using `notice` for this? It will make it stand out more.

File: index.js [hunk 2]
```diff
@@ -94,6 +96,11 @@ axios
     setOutput("id", response.data.id);
     setOutput("number", response.data.number);
     setOutput("state", response.data.state);
+    workFlowUrl =
+      "https://circleci.com/api/v2/pipeline/" + response.data.id + "/workflow";
```
Comment (whatisdot) [inline index.js]: You should use a string interpolation:
```ts
        workFlowUrl =`https://circleci.com/api/v2/pipeline/${response.data.id}/workflow`;
```

File: index.js [hunk 3]
```diff
@@ -102,3 +109,43 @@ axios
...
+const keepPolling = getInput("Follow").toLowerCase();
+
+const pollWorkflow = () => {
+  axios
+    .get(workFlowUrl, {
+      headers: headers,
+    })
+    .then((response) => {
+      if (response.data.items[0].status != "running") {
+        pipelineComplete = true;
```
Comment (whatisdot) [inline index.js]: According to [the CircleCI API](https://circleci.com/docs/api/v2/index.html#operation/getWorkflowById), possible statuses include:
> Enum: "success" "running" "not_run" "failed" "error" "failing" "on_hold" "canceled" "unauthorized" 

I found [this documentation for workflow states](https://github.com/circleci/circleci-docs/blob/master/jekyll/_cci2/workflows.md), which also lists a "needs setup" step not documented in the API... 🤔

Based on the two of these, we should continue polling when states are one of the following:
* `running`
* `not_run` (Workflow was never started)
* `on_hold` (A job in the workflow is waiting for approval)

You should check for any status within that set.

File: index.js [hunk 4]
```diff
@@ -102,3 +109,43 @@ axios
...
+        if (response.data.items[0].status == "success") {
+          info("CircleCI Pipeline is complete");
+        } else {
+          setFailed("CircleCI Pipeline failed");
```
Comment (whatisdot) [inline index.js]: You should add the workflow state to the log message.

File: index.js [hunk 5]
```diff
@@ -102,3 +109,43 @@ axios
...
+      startGroup("Failed to Poll CircleCI Pipeline");
+      coreError(error);
+      setFailed(error.message);
+      endGroup();
+    });
+};
+
+info("Poll CircleCI Workflow is is set to: " + keepPolling);
+
+if (keepPolling == "true") {
```
Comment (whatisdot) [inline index.js]: I would combine `pipelineComplete` with `keepPolling`, instead of managing two different flags.

**Review by whatisdot — DISMISSED** [review]

File: index.js [hunk 6]
```diff
@@ -102,3 +109,43 @@ axios
...
+    })
+    .catch((error) => {
+      startGroup("Failed to Poll CircleCI Workflow");
+      coreError(error);
+      setFailed(error.message);
+      endGroup();
+    });
+};
+
+if (followWorkflow) {
```
Comment (whatisdot) [inline index.js:146]: I think you only need to wrap the `info` call. The `setInterval` should immedately be unset if `followWorkflow` is false.

```typescript
if (followWorkflow) {
  info("Polling CircleCI Workflow");
}

const checkWebsiteStatus = setInterval(() => {
  if (!followWorkflow) {
    clearInterval(checkWebsiteStatus);
  } else {
    pollWorkflow();
  }
}, pollInterval);
```

File: index.js [hunk 7]
```diff
@@ -120,7 +120,11 @@ const pollWorkflow = () => {
       headers: headers,
     })
     .then((response) => {
-      if (response.data.items[0].status != "running") {
+      if (
+        response.data.items[0].status != "running" ||
```
Comment (whatisdot) [inline index.js]: I think you wanted boolean &&

"If the status isn't running, and isn't not_run, and isn't on_hold..."

I think you could also one-line this by using the `includes` method:
```typescript
if( ! ["not_run", "on_hold", "running"].includes(response.data.items[0].status) ) {
```

**Review by whatisdot — APPROVED** [review]
Tested this on an example project and it works as expected with both success and failure cases, along with polling and default non-polling behavior.

**Review by d10zero — APPROVED** [review]
Testing in VFE qa7 and works as expected. Nice !

**Review by whatisdot — APPROVED** [review]
Tested this and it works as expected. Now actions can wait until CircleCI finishes. Very nice :)

**Pull request merged**